flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesLeaderElector.java

```diff
@@ -0,0 +1,80 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.kubernetes.kubeclient.resources;
+
+import org.apache.flink.kubernetes.kubeclient.KubernetesLeaderElectionConfiguration;
+
+import io.fabric8.kubernetes.client.NamespacedKubernetesClient;
+import io.fabric8.kubernetes.client.extended.leaderelection.LeaderCallbacks;
+import io.fabric8.kubernetes.client.extended.leaderelection.LeaderElectionConfigBuilder;
+import io.fabric8.kubernetes.client.extended.leaderelection.LeaderElector;
+import io.fabric8.kubernetes.client.extended.leaderelection.resourcelock.ConfigMapLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.UUID;
+
+/**
+ * Represent {@link KubernetesLeaderElector} in kubernetes. {@link LeaderElector#run()} is a blocking call. It should be
+ *  run in the IO executor, not the main thread. The lifecycle is bound to single leader election. Once the leadership
+ * is revoked, as well as the {@link LeaderCallbackHandler#notLeader()} is called, the {@link LeaderElector#run()} will
+ * finish. To start another round of election, we need to trigger again.
+ */
+public class KubernetesLeaderElector extends LeaderElector<NamespacedKubernetesClient> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(KubernetesLeaderElector.class);
+	protected static final String LOCK_IDENTITY = UUID.randomUUID().toString();
+	protected static final String LEADER_ANNOTATION_KEY = "control-plane.alpha.kubernetes.io/leader";
+
+	public KubernetesLeaderElector(
+			NamespacedKubernetesClient kubernetesClient,
+			String namespace,
+			KubernetesLeaderElectionConfiguration leaderConfig,
+			LeaderCallbackHandler leaderCallbackHandler) {
+		super(kubernetesClient, new LeaderElectionConfigBuilder()
+			.withName(leaderConfig.getConfigMapName())
+			.withLeaseDuration(leaderConfig.getLeaseDuration())
+			.withLock(new ConfigMapLock(namespace, leaderConfig.getConfigMapName(), LOCK_IDENTITY))
+			.withRenewDeadline(leaderConfig.getRenewDeadline())
+			.withRetryPeriod(leaderConfig.getRetryPeriod())
+			.withLeaderCallbacks(new LeaderCallbacks(
+				leaderCallbackHandler::isLeader,
+				leaderCallbackHandler::notLeader,
+				newLeader -> LOG.info("New leader elected {}.", newLeader)
+			))
+			.build());
+		LOG.info("Create KubernetesLeaderElector {} with lock identity {}.",
+			leaderConfig.getConfigMapName(), LOCK_IDENTITY);
+	}
+
+	public boolean hasLeadership(KubernetesConfigMap configMap) {
+		final String leader = configMap.getAnnotations().get(LEADER_ANNOTATION_KEY);
```

Comment:
Could you explain in the JavaDocs, how the `LEADER_ANNOTATION_KEY` becomes an annotation of the `configMap`? And maybe add a bit of details which underlying K8s principles we are using here.

Reply:
I think your suggestion makes sense. I will add more description about where the annotation comes from and how it looks like.

```
  annotations:
    control-plane.alpha.kubernetes.io/leader: '{"holderIdentity":"623e39fb-70c3-44f1-811f-561ec4a28d75","leaseDuration":15.000000000,"acquireTime":"2020-10-20T04:06:31.431000Z","renewTime":"2020-10-21T13:42:46.843000Z","leaderTransitions":24195}'
```